src/include/storage/iceberg_delete.hpp

```diff
@@ -0,0 +1,137 @@
+//===----------------------------------------------------------------------===//
+//                         DuckDB
+//
+// storage/iceberg_delete.hpp
+//
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "duckdb/execution/physical_operator.hpp"
+#include "duckdb/planner/operator/logical_delete.hpp"
+#include "duckdb/common/multi_file/multi_file_reader.hpp"
+#include "storage/iceberg_metadata_info.hpp"
+#include "storage/iceberg_delete_filter.hpp"
+#include "storage/irc_table_entry.hpp"
+#include "storage/irc_schema_entry.hpp"
+
+namespace duckdb {
+
+struct IcebergDeleteMap {
+
+	optional_ptr<IcebergDeleteData> GetDeleteData(const string &filename) {
+		lock_guard<mutex> guard(lock);
+		auto entry = delete_data_map.find(filename);
+		if (entry == delete_data_map.end()) {
+			return nullptr;
+		}
+		return entry->second.get();
+	}
+
+private:
+	mutex lock;
+	unordered_map<string, shared_ptr<IcebergDeleteData>> delete_data_map;
+};
+
+struct WrittenColumnInfo {
+	WrittenColumnInfo() = default;
+	WrittenColumnInfo(LogicalType type_p, int32_t field_id) : type(std::move(type_p)), field_id(field_id) {
+	}
+
+	LogicalType type;
+	int32_t field_id;
+};
+
+class IcebergDeleteLocalState : public LocalSinkState {
+public:
+	string current_file_name;
+	vector<idx_t> file_row_numbers;
+};
+
+class IcebergDeleteGlobalState : public GlobalSinkState {
+public:
+	explicit IcebergDeleteGlobalState() {
+		written_columns["file_path"] = WrittenColumnInfo(LogicalType::VARCHAR, MultiFileReader::FILENAME_FIELD_ID);
+		written_columns["pos"] = WrittenColumnInfo(LogicalType::BIGINT, MultiFileReader::ORDINAL_FIELD_ID);
+	}
+
+	mutex lock;
+	unordered_map<string, IcebergDeleteFileInfo> written_files;
+	unordered_map<string, WrittenColumnInfo> written_columns;
+	idx_t total_deleted_count = 0;
+	unordered_map<string, vector<idx_t>> deleted_rows;
+	unordered_set<string> filenames;
+
+	void Flush(IcebergDeleteLocalState &local_state) {
+		auto &local_entry = local_state.file_row_numbers;
+		if (local_entry.empty()) {
+			return;
+		}
+		lock_guard<mutex> guard(lock);
+		auto &global_entry = deleted_rows[local_state.current_file_name];
+		global_entry.insert(global_entry.end(), local_entry.begin(), local_entry.end());
+		total_deleted_count += local_entry.size();
+		local_entry.clear();
+	}
+
+	void FinalFlush(IcebergDeleteLocalState &local_state) {
+		Flush(local_state);
+		// flush the file names to the global state
+		lock_guard<mutex> guard(lock);
+		filenames.emplace(local_state.current_file_name);
```

Comment:
Sorry to make yet another comment about this, but `filenames` can be removed from the global state entirely I think

Now I'm not sure the behavior is correct anymore, because we only write the "last" filename to the global state, but we do use `filenames` to verify, in `IcebergDelete::Finalize`

I'm pretty sure `filenames` is not necessary to have, the `deleted_rows` already serves this purpose